packages/amplify-cli/src/version-gating/index.ts

```diff
@@ -0,0 +1,202 @@
+import * as path from 'path';
+import _ from 'lodash';
+import semver from 'semver';
+import { $TSContext, JSONUtilities, stateManager } from 'amplify-cli-core';
+import * as CloudFormation from 'aws-sdk/clients/cloudformation';
+
+const packageJsonFileName = 'package.json';
+const disableVersionGatingEnvVarName = 'AMPLIFY_CLI_DISABLE_VERSION_CHECK';
+
+// Allowed Commands: 'configure', 'console', 'env list', 'help', 'init', 'logout', 'version', 'status', 'pull'
+
+// Type used internally for object matching
+type CommandMatch = {
+  plugins: string[];
+  command: string;
+};
+
+type Command = {
+  plugin?: string;
+  command: string;
+};
+
+type VersionGatingMetadata = {
+  DeployedByCLIVersion?: string;
+  MinimumCompatibleCLIVersion?: string;
+};
+
+export const versionGatingBlockedCommands: CommandMatch[] = [
```

Comment:
I think it would be better to have explicitly allowed command rather than blocked commands. For one, I think it will be a shorter list (is there anything besides `amplify version` and `amplify pull` we'd want to allow?). And two, it provides more security going forward if we add a command that should be version gated we don't have to remember to add it to this list

Reply:
it is 2 lists anyways that needs attention and maintain and since the functionality focuses on what is not allowed, that's why we use that for the list of commands.